2.train-VAE/betatcvae.py

```diff
@@ -0,0 +1,300 @@
+    fc2 : nn.Linear
+        Fully connected layer 2 in the decoder.
+    fc3 : nn.Linear
+        Fully connected layer 3 in the decoder.
+    decoder_input : nn.Linear
+        Input layer for the decoder.
+    decoder : nn.Sequential
+        Decoder network.
+    final_layer : nn.Linear
+        Final layer of the decoder.
+    
+    Methods:
+    --------
+    encode(input: torch.Tensor) -> List[torch.Tensor]
+        Encodes the input into the latent space.
+    decode(z: torch.Tensor) -> torch.Tensor
+        Decodes the latent variable into the reconstructed input.
+    reparameterize(mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor
+        Reparameterization trick to sample from the latent space.
+    forward(x: torch.Tensor)
+        Forward pass through the network.
+    log_density_gaussian(x: torch.Tensor, mu: torch.Tensor, logvar: torch.Tensor)
+        Computes the log density of a Gaussian.
+    loss_function(*args, **kwargs) -> dict
+        Computes the loss for the Beta-TCVAE.
+    """
+    
+    num_iter = 0  # Global static variable to keep track of iterations
+    
+    def __init__(self,
+                 input_dim: int,
+                 latent_dim: int,
+                 beta: float,
+                 hidden_dims: List = None,
+                 alpha: float = 1.0,
+                 gamma: float = 1.0,
+                 anneal_steps: int = 200,
+                 **kwargs) -> None:
+        super(BetaTCVAE, self).__init__()
+
+        # Initialize class attributes
+        self.latent_dim = latent_dim
+        self.anneal_steps = anneal_steps
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+
+        if hidden_dims is None:
+            hidden_dims = [256, 256, 256]  # Default hidden dimensions if none are provided
+
+        # Build Encoder
+        modules = []
+        in_dim = input_dim
+        for h_dim in hidden_dims:
+            modules.append(
+                nn.Sequential(
+                    nn.Linear(in_dim, h_dim),
+                    nn.ReLU())
+            )
+            in_dim = h_dim
+
+        self.encoder = nn.Sequential(*modules)
+        self.fc1 = nn.Linear(input_dim, 256) 
+        self.fc_mu = nn.Linear(256, latent_dim) 
+        self.fc_logvar = nn.Linear(256, latent_dim)  
+        self.fc2 = nn.Linear(latent_dim, 256)  
+        self.fc3 = nn.Linear(256, input_dim) 
+
+        # Build Decoder
+        self.decoder_input = nn.Linear(latent_dim, 256)  
+        modules = []
+        hidden_dims.reverse() 
+
+        for i in range(len(hidden_dims) - 1):
+            modules.append(
+                nn.Sequential(
+                    nn.Linear(hidden_dims[i], hidden_dims[i + 1]),
+                    nn.ReLU())
+            )
+
+        self.decoder = nn.Sequential(*modules)
+        self.final_layer = nn.Linear(hidden_dims[-1], input_dim)  
+
+    def encode(self, input: torch.Tensor) -> List[torch.Tensor]:
```

Comment:
consider adding docstrings to these methods

Reply:
done